### PR TITLE
ci: enable remote build caching for CI jobs

### PR DESCRIPTION
**Pull request by josephperrott**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the new behavior?
Enables remote caching for CI jobs.

This configuration:
- always reads from build cache on CI 
- only write to build cache for local builds for non-PR CI runs

### CODE REVIEW

File: .bazelrc [hunk 1]
```diff
@@ -138,9 +136,12 @@ build:remote --platforms=//tools:rbe_ubuntu1604-angular
 # Remote instance.
 build:remote --remote_instance_name=projects/internal-200822/instances/default_instance
 
-# Do not accept remote cache.
-# We need to understand the security risks of using prior build artifacts.
+# Remote cache
+build:remote --remote_http_cache=https://storage.googleapis.com/angular-team-cache
+# By default, do not accept remote cache, to be enabled on for CI based on environment
```
Comment (gkalpak) [inline .bazelrc]: _enabled on_ --> _enabled_/_turned on_ (?)

File: .bazelrc [hunk 2]
```diff
@@ -138,9 +136,12 @@ build:remote --platforms=//tools:rbe_ubuntu1604-angular
...
 build:remote --remote_accept_cached=false
+# By default, do not upload local results to cache, to be enabled on for CI based on environment
```
Comment (gkalpak) [inline .bazelrc]: _enabled on_ --> _enabled_/_turned on_ (?)

File: .circleci/config.yml [hunk 3]
```diff
@@ -58,7 +58,20 @@ var_5: &setup_bazel_remote_execution
       # cause decryption failures based on the openssl version. https://stackoverflow.com/a/39641378/4317734
       openssl aes-256-cbc -d -in .circleci/gcp_token -md md5 -k "$CI_REPO_NAME" -out /home/circleci/.gcp_credentials
       echo "export GOOGLE_APPLICATION_CREDENTIALS=/home/circleci/.gcp_credentials" >> $BASH_ENV
-      sudo bash -c "echo 'build --config=remote' >> /etc/bazel.bazelrc"
+      sudo bash -c "echo -e 'build --config=remote\n' >> /etc/bazel.bazelrc"
+  run:
+    name: Setting up remote caching configuration for CI
+    command: >
+      sudo bash -c "echo 'build:remote --remote_accept_cached=true\n' >> /etc/bazel.bazelrc"
```
Comment (gkalpak) [inline .circleci/config.yml]: `\n` should be used with `echo -e` :wink:

File: .bazelrc [hunk 4]
```diff
@@ -1,8 +1,6 @@
 ################################
 # Settings for Angular team members only
```
Comment (alexeagle) [inline .bazelrc]: no more settings that are specific for us?

Comment (gkalpak) [inline .bazelrc]: Is this "header" still relevant? (I can't tell 😟)

**Review by alexeagle — COMMENTED** [review]
hey @huangw5 we could use your input to make sure we are doing this right

File: .circleci/config.yml [hunk 5]
```diff
@@ -58,7 +58,20 @@ var_5: &setup_bazel_remote_execution
...
-      sudo bash -c "echo 'build --config=remote' >> /etc/bazel.bazelrc"
+      sudo bash -c "echo -e 'build --config=remote\n' >> /etc/bazel.bazelrc"
+  run:
+    name: Setting up remote caching configuration for CI
+    command: |
+      sudo bash -c "echo 'build:remote --remote_accept_cached=true\n' >> /etc/bazel.bazelrc"
+      echo "Reading from remote cache for bazel remote jobs."
+      if [[ -n "${CIRCLE_PR_NUMBER}" ]]; then
+        # The current CI run from a commit to a PR
+        sudo bash -c "echo 'build:remote --remote_upload_local_results=false\n' >> /etc/bazel.bazelrc"
```
Comment (gkalpak) [inline .circleci/config.yml]: `\n` should be used with `echo -e` :wink:

File: .bazelrc (hunk 4 above)
Comment (josephperrott) [reply .bazelrc]: Not at this point.  Because we always want to have the cache url defined for CI

Comment (josephperrott) [reply .bazelrc]: I suppose we can just remove it, if we end up having angular team member specific settings then we can add it back in.

Comment (josephperrott) [reply .bazelrc]: Resolved in another comment.

**Comment by huangw5** [comment]
> hey @huangw5 we could use your input to make sure we are doing this right

This looks good to me. Sorry I missed the Github notifications.

**Comment by josephperrott** [comment]
@gkalpak Can you take another look when you have a chance?

File: .circleci/config.yml [hunk 6]
```diff
@@ -58,7 +58,18 @@ var_5: &setup_bazel_remote_execution
       # cause decryption failures based on the openssl version. https://stackoverflow.com/a/39641378/4317734
       openssl aes-256-cbc -d -in .circleci/gcp_token -md md5 -k "$CI_REPO_NAME" -out /home/circleci/.gcp_credentials
       echo "export GOOGLE_APPLICATION_CREDENTIALS=/home/circleci/.gcp_credentials" >> $BASH_ENV
-      sudo bash -c "echo 'build --config=remote' >> /etc/bazel.bazelrc"
+      touch .bazelrc.user
+      sudo bash -c "echo -e 'build --config=remote\n' >> .bazelrc.user"
+      sudo bash -c "echo -e 'build:remote --remote_accept_cached=true\n' >> .bazelrc.user"
+      echo "Reading from remote cache for bazel remote jobs."
+      if [[ -n "${CIRCLE_PR_NUMBER}" ]]; then
```
Comment (IgorMinar) [inline .circleci/config.yml]: please use CI_PULL_REQUEST instead - we are trying to not use CI provider specific vars to increase portability and readability (see: https://github.com/angular/angular/blob/a5dd4edab92b10c093596d554d07cba93f5f8c76/.circleci/env.sh#L32)

Comment (gkalpak) [reply .circleci/config.yml]: +1

Generally, my rule has been to never use CI-specific vars outside of `.circleci/`, but accept CI-specific ones inside `.circleci/` (which is CircleCI-specific by definition).
But 100% agree that we should try to use CI-agnostic vars even inside `.circleci/`.

Comment (gkalpak) [reply .circleci/config.yml]: BTW, switching to `CI_PULL_REQUEST` would also require changing the `-n` operator (e.g. `"$CI_PULL_REQUEST" == "false"`).

**Review by gkalpak — APPROVED** [review]
🎉 LGTM (pending [Igor's comment][1])

[1]: https://github.com/angular/angular/pull/31204/commits#r296959010

**Pull request closed without merging**